src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -117959,6 +117959,161 @@ static void M1(C c, string? s)
                     Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(17, 17));
         }
 
+        [Fact, WorkItem(40577, "https://github.com/dotnet/roslyn/issues/40577")]
+        public void IEquatableEqualsMethod_ImplInBaseType()
+        {
+            var source = @"
+using System;
+
+#nullable enable
+class B : IEquatable<string>
+{
+    public bool Equals(string? s) => throw null!;
+}
+
+class C : B
+{
+    static void M1(C c, string? s)
+    {
+        if (c.Equals(s))
+        {
+            _ = s.ToString();
+        }
+        else
+        {
+            _ = s.ToString(); // 1
+        }
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                    // (20,17): warning CS8602: Dereference of a possibly null reference.
+                    //             _ = s.ToString(); // 1
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(20, 17));
+        }
+
+        [Fact, WorkItem(40577, "https://github.com/dotnet/roslyn/issues/40577")]
+        public void IEquatableEqualsMethod_ExplicitImpl()
+        {
+            var source = @"
+using System;
+
+#nullable enable
+class C : IEquatable<string>
+{
+    public bool Equals(string? s) => throw null!;
+    bool IEquatable<string>.Equals(string? s) => throw null!;
+
+    static void M1(C c, string? s)
+    {
+        _ = c.Equals(s)
+            ? s.ToString() // 1
+            : s.ToString(); // 2
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (13,15): warning CS8602: Dereference of a possibly null reference.
+                //             ? s.ToString() // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(13, 15),
+                // (14,15): warning CS8602: Dereference of a possibly null reference.
+                //             : s.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(14, 15));
+        }
+
+        [Fact, WorkItem(40577, "https://github.com/dotnet/roslyn/issues/40577")]
+        public void IEquatableEqualsMethod_ExplicitImplInDerivedType()
+        {
+            var source = @"
+using System;
+
+#nullable enable
+class B : IEquatable<string>
```

Comment:
>IEquatable<string> [](start = 10, length = 18)

It feels like we are missing a test for scenario when B doesn't implement the interface, but has the method.